File: 4/sol_part1.jl
function main()
    # Check if a file was provided as an argument
    if length(ARGS) != 1
        println("Please provide exactly one file as a parameter.")
        return
    end

    # Get the filename from the arguments
    filename = ARGS[1]
    file_content = ""

    # Try to open the file and read its contents
    try
        file_content = read(filename, String)
    catch e
        println("Error reading file: $e")
        return
    end

    lines = split(file_content, '\n')
    grid = [collect(line) for line in lines]

    counter = 0

    for i in 1:size(grid, 1)
        for j in 1:length(grid[i])
            if grid[i][j] == 'X'
                # Top-left diagonal
                if i >= 4 && j >= 4 && 
                   grid[i-1][j-1] == 'M' && grid[i-2][j-2] == 'A' && grid[i-3][j-3] == 'S'
                    counter += 1
                end

                # Top
                if i >= 4 && 
                   grid[i-1][j] == 'M' && grid[i-2][j] == 'A' && grid[i-3][j] == 'S'
                    counter += 1
                end

                # Top-right diagonal
                if i >= 4 && j <= length(grid[i]) - 3 && 
                   grid[i-1][j+1] == 'M' && grid[i-2][j+2] == 'A' && grid[i-3][j+3] == 'S'
                    counter += 1
                end

                # Left
                if j >= 4 && 
                   grid[i][j-1] == 'M' && grid[i][j-2] == 'A' && grid[i][j-3] == 'S'
                    counter += 1
                end

                # Right
                if j <= length(grid[i]) - 3 && 
                   grid[i][j+1] == 'M' && grid[i][j+2] == 'A' && grid[i][j+3] == 'S'
                    counter += 1
                end

                # Bottom-left diagonal
                if i <= size(grid, 1) - 3 && j >= 4 && 
                   grid[i+1][j-1] == 'M' && grid[i+2][j-2] == 'A' && grid[i+3][j-3] == 'S'
                    counter += 1
                end

                # Bottom
                if i <= size(grid, 1) - 3 && 
                   grid[i+1][j] == 'M' && grid[i+2][j] == 'A' && grid[i+3][j] == 'S'
                    counter += 1
                end

                # Bottom-right diagonal
                if i <= size(grid, 1) - 3 && j <= length(grid[i]) - 3 && 
                   grid[i+1][j+1] == 'M' && grid[i+2][j+2] == 'A' && grid[i+3][j+3] == 'S'
                    counter += 1
                end
            end
        end
    end

    println(counter)
end

main()

File: 6/sol_part1.js
const fs = require('fs');

function convertToGrid(input) {
    return input.split('\n').map(line => line.split(''));
}

function findGuard(grid) {    
    const specialChars = ['^', '>', '<', 'v'];
    
    for (let row = 0; row < grid.length; row++) {
        for (let col = 0; col < grid[row].length; col++) {
            if (specialChars.includes(grid[row][col])) {
                return { row, col, char: grid[row][col] };
            }
        }
    }
    
    // Return null if not found
    return null;
}

function execute(grid, guard) {
    let outside = false;
    let prevStep = 0; // 1 -> row++, 2 -> row--, 3 -> col++, 4 -> col--

    while (!outside) {
        if (guard.row >= grid.length || guard.row < 0 || guard.col >= grid[guard.row].length || guard.col < 0) {
            outside = true;
        } else if (grid[guard.row][guard.col] == "#") {
            switch (prevStep) {
                case 1:
                    guard.char = ">";
                    guard.row++;
                    break;
                case 2:
                    guard.char = "<";
                    guard.row--;
                    break;
                case 3:
                    guard.char = "v";
                    guard.col--;
                    break;
                case 4:
                    guard.char = "^";
                    guard.col++;
                    break;
                default:
                    console.log("Error: invalid state.") 
            }
        } else {
            if (grid[guard.row][guard.col] == ".") {
                grid[guard.row][guard.col] = "X";
            }

            switch (guard.char) {
                case "^":
                    guard.row--;
                    prevStep = 1;
                    break;
                case ">":
                    guard.col++;
                    prevStep = 3;
                    break;
                case "<":
                    guard.col--;
                    prevStep = 4;
                    break;
                case "v":
                    guard.row++;
                    prevStep = 2;
                    break;
                default:
                    console.log("Error: invalid state.") 
            }
        }
    }

    return grid
}

function part1(grid, guard) {
    grid = execute(grid, guard);

    let counter = 1; // For the initial position.

    for (let row = 0; row < grid.length; row++) {
        for (let col = 0; col < grid[row].length; col++) {
            if (grid[row][col] == "X") {
                counter++;
            }
        }
    }

    return counter;
}

function main() {
    if (process.argv.length !== 3) {
        console.log('Please provide exactly one file as a parameter.');
        return;
    }

    const filePath = process.argv[2];
    let fileContent = "";

    try {
        fileContent = fs.readFileSync(filePath, 'utf-8');
        // console.log(fileContent);
    } catch (error) {
        console.error(`Error reading file: ${error.message}`);
    }

    const grid = convertToGrid(fileContent.trim());
    const guard = findGuard(grid);

    result = part1(grid, guard);
    console.log(result);
}

main();
